compiler/src/dotty/tools/dotc/core/Denotations.scala

```diff
@@ -1058,7 +1063,10 @@ object Denotations {
         def needsPrefix =
           // For opaque types, the prefix is used in `ElimOpaques#transform`,
           // without this i7159.scala would fail when compiled from tasty.
-          symbol.is(Opaque)
+          symbol.is(Opaque) ||
+          // SingleDenotation#signature relies on the prefix to compute
+          // an appropriate signature for methods.
+          info.isInstanceOf[MethodOrPoly] && symbol.is(JavaDefined) != pre.classSymbol.is(JavaDefined)
```

Comment:
Why does the prefix `isJavaDefined` status override the symbol's status?

Reply:
When checking if `foo` in `B` overrides `foo` in `A`, we look at the symbol of `foo` in `A` as seen from prefix `B.this` and compare their signature, so we need to rely on this prefix to know what signature to use to match the signature of the override at its definition site.